cms/envs/common.py

```diff
@@ -352,18 +352,6 @@
     'ENABLE_CHANGE_USER_PASSWORD_ADMIN': False,
 
     ### ORA Feature Flags ###
-
-    # .. toggle_name: ENABLE_ORA_TEAM_SUBMISSIONS
-    # .. toggle_implementation: DjangoSetting
-    # .. toggle_default: False
-    # .. toggle_description: Set to True to enable team-based ORA submissions.
-    # .. toggle_use_cases: temporary
-    # .. toggle_creation_date: 2020-03-03
-    # .. toggle_target_removal_date: None
-    # .. toggle_tickets: https://openedx.atlassian.net/browse/EDUCATOR-4951
-    # .. toggle_warnings: This temporary feature toggle does not have a target removal date.
-    'ENABLE_ORA_TEAM_SUBMISSIONS': False,
```

Comment:
> Would it be possible for this code to use the SettingToggle and SettingDictToggle class instances to write the setting with its default value if it doesn't yet exist in settings?

Do you mean using SettingDictToggle from within cms/envs/common.py? I don't think that would be possible, as it would introduce a circular dependency.

> If so, replacing calls to settings.FEATURES.get('ENTRANCE_EXAMS') (or any setting) may not be required.

I think it's better if we do remove them as much as we can. Can I suggest that we keep the SettingDictToggle as it is, and also keep the default value from the common.py file? It looks like a good compromise to me.

> We were also wondering if having to replace all the reads of the setting slowed you down a lot, or would slow down a community effort to annotate these properly?

No it's not a problem, as long as it's in just one repository and I can handle all the changes in a single PR.

Reply:
Some other possibilities and questions:
* Start implementing defaults.yml.
* We could allow SettingToggle to accept None as a default. The implementer can decide whether it is required or not. If default value were optional, we’d need to choose between False and None as the default, and None might be a cleaner interface.
* Is there a way for the linter to access defaults.yml should we want to ensure the default in the annotation? 
* A question for later: if an annotation/SettingToggle in a library repo declares a specific default, but a specific IDA needs to override the default in its defaults.py, how do we get the readthedocs report to not show the wrong default?

Reply:
> Start implementing defaults.yml.

I think we should get this merged before starting to think of bigger, over-arching changes :)

> We could allow SettingToggle to accept None as a default. The implementer can decide whether it is required or not. If default value were optional, we’d need to choose between False and None as the default, and None might be a cleaner interface.

My first reaction to this would be a big no no no... But I don't understand how that would work in practice.

> A question for later: if an annotation/SettingToggle in a library repo declares a specific default, but a specific IDA needs to override the default in its defaults.py, how do we get the readthedocs report to not show the wrong default?

If IDA 1 needs to override a setting from IDA 2, then it's doing this wrong :) It is the responsibility of the operator of IDA 2 to define the setting accurately.

Reply:
Imagine a toggle defined in edx-drf-extensions with a default and annotation. Imagine IDA1 uses that default, and IDA2 flips that default In its own defaults.

Reply:
This is a problem of code duplication. If we have:

```
toggle = SetttingToggle("MYSETTING", default=True)
```

and in settings/common.py:

```
MYSETTING = False
```

then having both values kept in sync requires manual work. We can remove this code duplication, at the cost of backward incompatibility. I don't think we can have both backward compatibility and no code duplication.

Reply:
I’m looking for ways to get where we want to go and to reduce risk when we might have various community members documenting and moving lots of settings.

One possibility is to punt by adding a comment here that points to where the setting is defined/annotated. This would help us track what was annotated and enable us to return to this in the future.

Other possibilities include:
- patching settings to see if they are requested outside and inside a SettingToggle. (Probably too costly and risky.)
- move these (now or later) to defaults.yml and allow the toggle state endpoint to see these defined defaults and ensure they match those defined in SettingToggles.
- ?

Reply:
With (low) 65% confidence, my current option would be to introduce defaults.yml, have duplication, have low risk of breaking things, and a reasonably simple ability to ensure that the duplication is consistent.

Reply:
I'd like to suggest an alternative approach, where the default SettingDictToggle value is more explicitely linked to the setting, if it exists. We could implement a `SafeSettingDictToggle` which would look like this:

```
class SafeSettingDictToggle(SettingDictToggle):
    def is_enabled(self):
        setting_dict = getattr(settings, self.name, {})
        setting_value = setting_dict.get(self.key))
        if setting_value is not None:
            if setting_value != self.default:
                raise ArgumentError("Incoherent setting {}['{}']: found value {} with toggle default value is".format(self.name, self.key, setting_value, self.default))
        return self.default
```

This would guarantee that the SettingDictToggle default value was correctly defined at runtime. Once we migrate all settings to SafeSettingToggle/SafeSettingDictToggle we could then switch to SettingToggle/SettingDictToggle instances.

Reply:
We agreed in another venue not to try to implement the `Safe` option described above, because there are legitimate reasons why the default may be overridden. We also agreed to clean up defaults where possible, as described in the ADR, and rely on our ability to adjust settings quickly should this cause an issue.